src/services/safe_border.py

```diff
@@ -0,0 +1,164 @@
+
+class SafeBorder:
+    def __init__(self, w3: Web3) -> None:
+        self.w3 = w3
+        self.lido_contracts = w3.lido_contracts
+
+    def get_safe_border_epoch(self, is_bunker: bool, blockstamp: BlockStamp, chain_config: ChainConfig):
+        if not is_bunker:
+            return self._get_new_requests_border_epoch(blockstamp)
+
+        self.chain_config = chain_config
+        
+        negative_rebase_border_epoch = self._get_negative_rebase_border_epoch(blockstamp)
+        associated_slashings_border_epoch = self._get_associated_slashings_border_epoch(blockstamp)
+
+        return min(
+            negative_rebase_border_epoch,
+            associated_slashings_border_epoch
+        )
+
+    def _get_new_requests_border_epoch(self, blockstamp: BlockStamp):
+        return self.get_epoch_by_slot(blockstamp.ref_slot) - DEFAULT_SHIFT
+
+    def _get_negative_rebase_border_epoch(self, blockstamp: BlockStamp):  
+        bunker_start_timestamp = self._get_bunker_mode_start_timestamp(blockstamp)
+        bunker_start_slot = self.get_slot_by_timestamp(bunker_start_timestamp)
+
+        last_withdrawable_epoch = self.get_epoch_by_slot(bunker_start_slot)
+
+        if bunker_start_slot > blockstamp.ref_slot:
+            last_withdrawable_epoch = self.get_epoch_by_slot(self._get_last_successful_report_slot(blockstamp))
+
+        latest_allowable_epoch = last_withdrawable_epoch - DEFAULT_SHIFT
+        earliest_allowable_epoch = self.get_epoch_by_slot(blockstamp.ref_slot) - MAX_NEGATIVE_REBASE_SHIFT
+
+        return max(earliest_allowable_epoch, latest_allowable_epoch)
+
+    def _get_associated_slashings_border_epoch(self, blockstamp: BlockStamp):
+        earliest_slashed_epoch = self._get_earliest_slashed_epoch_among_incomplete_slashings(blockstamp)
+
+        if earliest_slashed_epoch is None:
+            return self.get_epoch_by_slot(blockstamp.ref_slot) - DEFAULT_SHIFT
+        
+        return earliest_slashed_epoch - DEFAULT_SHIFT
+
+    def _get_earliest_slashed_epoch_among_incomplete_slashings(self, blockstamp: BlockStamp):
+        validators = self._get_lido_validators(blockstamp)       
+        validators_slashed = filter_slashed_validators(validators)
+
+        # Here we filter not by exit_epoch but by withdrawable_epoch because exited operators can still be slashed.
+        # See more here https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#helpers
+        # at `get_eligible_validator_indices` method.
+        validators_slashed_non_withdrawable = filter_non_withdrawable_validators(validators_slashed, blockstamp.ref_slot)
+
+        if len(validators_slashed_non_withdrawable) == 0:
+            return None
+
+        validators_with_earliest_exit_epoch = self._filter_validators_with_earliest_exit_epoch(validators_slashed_non_withdrawable)
+        earliest_slashed_epoch_predicted = self._predict_earliest_slashed_epoch(validators_with_earliest_exit_epoch[0])
```

Comment:
If at least one of these validators would have predictable epoch, there is no need to find it with binary search
Maybe check all of them, not only the first one?